cumulusci/tasks/salesforce/package_upload.py

```diff
@@ -0,0 +1,222 @@
+
+        self._make_package_upload_request(package_info, package_id)
+
+        if self.upload["Status"] == "ERROR":
+            self._log_package_upload_errors()
+        else:
+            self._set_package_version_values_on_self()
+            self._set_return_values(package_id)
+            self._log_package_upload_success()
+
+    def _get_package_id_and_info(self):
+        namespace = self.options["namespace"]
+        package = self._get_one_record(
+            f"SELECT Id FROM MetadataPackage WHERE NamespacePrefix='{namespace}'",
+            f"No package found with namespace {namespace}",
+        )
+        package_id = package["Id"]
+        production = self.options.get("production", False) in [True, "True", "true"]
+
+        package_info = {
+            "VersionName": self.options["name"],
+            "IsReleaseVersion": production,
+            "MetadataPackageId": package_id,
+        }
+        self._set_package_info_values_from_options(package_info)
+
+        return package_id, package_info
+
+    def _set_package_info_values_from_options(self, package_info):
+        if "description" in self.options:
+            package_info["Description"] = self.options["description"]
+        if "password" in self.options:
+            package_info["Password"] = self.options["password"]
+        if "post_install_url" in self.options:
+            package_info["PostInstallUrl"] = self.options["post_install_url"]
+        if "release_notes_url" in self.options:
+            package_info["ReleaseNotesUrl"] = self.options["release_notes_url"]
+
+    def _make_package_upload_request(self, package_info, package_id):
+        """Creates a PackageUploadRequest in self.upload"""
+        PackageUploadRequest = self._get_tooling_object("PackageUploadRequest")
+
+        self._upload_start_time = time.time()
+        self.upload = PackageUploadRequest.create(package_info)
+
+        self.upload_id = self.upload["id"]
+        self.logger.info(
+            f"Created PackageUploadRequest {self.upload_id} for Package {package_id}"
+        )
+        self._poll()
+
+    def _log_package_upload_errors(self):
+        self.logger.error("Package upload failed with the following errors")
+        error = {"message": ""}
+        apex_test_failures = False
+        for error in self.upload["Errors"]["errors"]:
+            self.logger.error("  {}".format(error["message"]))
+            if error["message"] == "ApexTestFailure":
+                apex_test_failures = True
+
+        if apex_test_failures:
+            self._log_apex_test_failures()
+
+        exception = self._get_exception_type(error)
+        raise exception("Package upload failed")
+
+    def _log_apex_test_failures(self):
+        self.logger.error("Failed Apex Tests:")
+        soql_query = self._get_failed_tests_soql_query()
+        results = self.tooling.query(soql_query)
+
+        table_header_row = ["Class", "Method", "Stacktrace"]
```

Comment:
Would 'Apex Class' be better? Title of the table is 'Apex Test Failures' though...

Reply:
I think just "Class" is fine, given that context.